File: jfingerprintmanager/src/main/java/com/jesusm/jfingerprintmanager/base/FingerprintBaseDialogPresenter.java
package com.jesusm.jfingerprintmanager.base;


import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.hardware.fingerprint.FingerprintManagerCompat;

import com.jesusm.jfingerprintmanager.JFingerprintManager;
import com.jesusm.jfingerprintmanager.base.hardware.FingerprintHardware;
import com.jesusm.jfingerprintmanager.base.model.FingerprintManagerCancellationSignal;

import javax.crypto.Cipher;

import static com.jesusm.jfingerprintmanager.base.FingerprintBaseDialogPresenter.BaseStage.FINGERPRINT;

public abstract class FingerprintBaseDialogPresenter extends
        FingerprintManagerCompat.AuthenticationCallback {

    private FingerprintManagerCancellationSignal cancellationSignal;
    protected FingerprintHardware fingerprintHardware;
    @Nullable
    private Cipher defaultCipher;

    public interface Stage {
        String id();
    }

    /**
     * Enumeration to indicate which authentication method the user is trying to authenticate with.
     */
    public enum BaseStage implements Stage {
        FINGERPRINT() {
            @Override
            public String id() {
                return "fingerprint";
            }
        }
    }

    protected Stage stage = FINGERPRINT;
    @NonNull
    protected View view;

    public FingerprintBaseDialogPresenter(@NonNull View view) {
        this.view = view;
        cancellationSignal = new FingerprintManagerCancellationSignal();
    }

    public void pause() {
        cancelFingerprintAuthenticationListener();
    }

    protected void cancelFingerprintAuthenticationListener() {
        cancellationSignal.cancel();
    }

    public void onViewShown() {
        updateStage();
    }

    @CallSuper
    protected void updateStage() {
        if (stage.id().equals(FINGERPRINT.id())) {
            displayFingerprint();
        }
    }

    private void displayFingerprint() {
        view.onFingerprintDisplayed();
        startAuthenticationListener();
    }

    public void setStage(Stage stage) {
        this.stage = stage;
    }

    public Stage getStage() {
        return stage;
    }

    public void setFingerprintHardware(@NonNull FingerprintHardware fingerprintHardware,
                                       @Nullable Cipher defaultCipher) {
        this.fingerprintHardware = fingerprintHardware;
        this.defaultCipher = defaultCipher;
    }

    @SuppressWarnings("MissingPermission")
    private void startAuthenticationListener() {
        cancellationSignal.start();

        // As soon as this is called, we are listening for fingerprint introduction.
        FingerprintManagerCompat.CryptoObject cryptoObject = new FingerprintManagerCompat.CryptoObject(defaultCipher);
        fingerprintHardware.authenticate(cryptoObject, 0, cancellationSignal.getCancellationSignal(), this, null);
    }

    public void setCancellationSignal(FingerprintManagerCancellationSignal cancellationSignal) {
        this.cancellationSignal = cancellationSignal;
    }

    @Override
    public void onAuthenticationError(int errMsgId, CharSequence errString) {
        if (!cancellationSignal.isCancelled()) {
            view.onFingerprintNotRecognized();
            close();
        }
    }

    @Override
    public void onAuthenticationHelp(int helpMsgId, CharSequence helpString) {
        view.onAuthenticationFailedWithHelp(helpString.toString());
        close();
    }

    @Override
    public void onAuthenticationFailed() {
        view.onFingerprintNotRecognized();
        close();
    }

    protected void close() {
        view.close();
    }

    public interface View extends JFingerprintManager.FingerprintBaseCallback {
        void onFingerprintDisplayed();

        void close();
    }
}

File: jfingerprintmanager/src/main/kotlin/com/jesusm/jfingerprintmanager/base/ui/presenter/FingerprintBaseDialogPresenter.kt
package com.jesusm.jfingerprintmanager.base.ui.presenter

import android.support.v4.hardware.fingerprint.FingerprintManagerCompat
import com.jesusm.jfingerprintmanager.JFingerprintManager
import com.jesusm.jfingerprintmanager.base.hardware.FingerprintHardware
import com.jesusm.jfingerprintmanager.base.model.FingerprintManagerCancellationSignal

abstract class FingerprintBaseDialogPresenter(val view: View, var stage: Stage = FingerprintBaseDialogPresenter.Stage.FINGERPRINT,
                                              var cancellationSignal: FingerprintManagerCancellationSignal = FingerprintManagerCancellationSignal(),
                                              var fingerprintHardware: FingerprintHardware? = null,
                                              var cryptoObject: FingerprintManagerCompat.CryptoObject? = null) : FingerprintManagerCompat.AuthenticationCallback() {
    /**
     * Enumeration to indicate which authentication method the user is trying to authenticate with.
     */
    enum class Stage(val id: String) {
        FINGERPRINT("fingerprint"),
        NEW_FINGERPRINT_ENROLLED("newFingerprintEnrolled"),
        PASSWORD("password")
    }

    fun pause() {
        cancelFingerprintAuthenticationListener()
    }

    fun close() {
        view.close()
    }

    fun cancelFingerprintAuthenticationListener() {
        cancellationSignal.cancel()
    }

    open fun onViewShown() {
        updateStage()
    }

    open fun updateStage() {
        if (stage.id == Stage.FINGERPRINT.id) {
            displayFingerprint()
        }
    }

    private fun displayFingerprint() {
        view.onFingerprintDisplayed()
        startAuthenticationListener()
    }

    private fun startAuthenticationListener() {
        cancellationSignal.start()

        // As soon as this is called, we are listening for fingerprint introduction.
        cancellationSignal.cancellationSignal?.let {
            fingerprintHardware?.authenticate(cryptoObject, 0, it, this, null)
        }
    }

    fun setFingerprintHardware(hardware: FingerprintHardware, cryptoObject: FingerprintManagerCompat.CryptoObject?) {
        fingerprintHardware = hardware
        this.cryptoObject = cryptoObject
    }

    override fun onAuthenticationError(errMsgId: Int, errString: CharSequence?) {
        if (cancellationSignal.isCancelled().not()) {
            onAuthenticationFailed()
        }
    }

    override fun onAuthenticationHelp(helpMsgId: Int, helpString: CharSequence?) {
        view.onAuthenticationFailedWithHelp(helpString.toString())
        close()
    }

    override fun onAuthenticationFailed() {
        view.onFingerprintNotRecognized()
        close()
    }

    fun onDialogCancelled() =
            view.onCancelled()

    interface View : JFingerprintManager.FingerprintBaseCallback {
        fun onFingerprintDisplayed()

        fun close()
    }
}